hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerImpl.java

```diff
@@ -341,6 +341,9 @@ protected void closeContainersForPipeline(final PipelineID pipelineId)
   public void closePipeline(Pipeline pipeline, boolean onTimeout)
       throws IOException {
     PipelineID pipelineID = pipeline.getId();
+    // close containers.
+    closeContainersForPipeline(pipelineID);
```

Comment:
Current logic of closeContainersForpipeline is just firing events, so lets say events have not completed and pipeline closed or removed we will be in situation like containers open with out a pipeline.

So, I think we should update closeContainersForPipeline as below. 
With this approach we are making sure containers are in closing before pipeline remove/close log in ratis.
```
  protected void closeContainersForPipeline(final PipelineID pipelineId)
      throws IOException {
    Set<ContainerID> containerIDs = stateManager.getContainers(pipelineId);
    for (ContainerID containerID : containerIDs) {
      if (scmContext.getScm().getContainerManager()
          .getContainer(containerID).getState()
          == HddsProtos.LifeCycleState.OPEN) {
        try {
          scmContext.getScm().getContainerManager().updateContainerState(
              containerID, HddsProtos.LifeCycleEvent.FINALIZE);
        } catch (InvalidNodeStateException ex) {
          throw new IOException(ex);
        }
      }

      eventPublisher.fireEvent(SCMEvents.CLOSE_CONTAINER, containerID);
    }
  }
```

Reply:
+1 to the above suggestion. We cannot rely on async handling of moving container from OPEN to CLOSING.

Reply:
Thank you for the review and suggestions. I will update the PR with this change.

Reply:
+1 n the idea

Reply:
+1, the idea looks good！